invenio_rdm_records/services/community_inclusion/service.py

```diff
@@ -0,0 +1,94 @@
+# it under the terms of the MIT License; see LICENSE file for more details.
+
+"""Community Inclusion Service."""
+
+from flask import current_app
+from invenio_communities import current_communities
+from invenio_communities.communities.records.systemfields.access import (
+    CommunityAccess,
+    VisibilityEnum,
+)
+from invenio_i18n import lazy_gettext as _
+from invenio_records_resources.services.errors import PermissionDeniedError
+from invenio_requests import current_requests_service
+
+from invenio_rdm_records.records.systemfields.access.field.record import (
+    AccessStatusEnum,
+)
+from invenio_rdm_records.requests.community_inclusion import CommunityInclusion
+from invenio_rdm_records.requests.community_submission import CommunitySubmission
+from invenio_rdm_records.services.errors import (
+    CommunityInclusionInconsistentAccessRestrictions,
+)
+
+
+class CommunityInclusionService:
+    """Service for including a record in a community.
+
+    The RDM Requests service wraps some operations of the generic requests service,
+    implementing RDM business logic.
+
+    Note: this service is meant to be used by other services, and not by a resource.
+    The uow should be passed by the caller.
+    """
+
+    @property
+    def supported_types(self):
+        """Supported request types."""
+        types = {CommunitySubmission.type_id, CommunityInclusion.type_id}
+        types.update(current_app.config.get("RDM_RECORDS_REVIEWS", []))
+        return types
+
+    def submit(self, identity, record, community, request, data, uow):
+        """Submit a request to include a record in a community.
+
+        It ensures that public records cannot be included in restricted communities.
+        """
+        if request.type.type_id not in self.supported_types:
+            raise ValueError("Invalid request type.")
+
+        assert CommunityAccess.validate_visibility_level(VisibilityEnum.RESTRICTED)
+        community_is_restricted = community["access"]["visibility"] == str(
+            VisibilityEnum.RESTRICTED
+        )
+
+        record_is_restricted = record.access.status == AccessStatusEnum.RESTRICTED
+
+        if community_is_restricted and not record_is_restricted:
+            raise CommunityInclusionInconsistentAccessRestrictions()
+
+        # All other preconditions can be checked by the action itself which can
+        # raise appropriate exceptions.
+        return current_requests_service.execute_action(
+            identity, request.id, "submit", data=data, uow=uow
+        )
+
+    def include(self, identity, community, request, uow):
+        """Accept the request to include the record in the community.
+
+        Request will be accepted based on community policy and identity permissions
+        """
+        if request.type in self.supported_types:
+            raise ValueError("Invalid request type.")
+
+        try:
+            current_communities.service.require_permission(
```

Comment:
there was also check_permission method if you want to avoid try catch

Reply:
Thanks, replaced